## SQL Project - Goals
the next project will analyze a database of a books' app in order to gain knowledge about users' behavior to propose a value proposition for a new product.
THe dataset includes information about books, publishers, authors and user reviews.

## Importing database

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])


engine = create_engine(connection_string, connect_args={'sslmode':'require'})


#### Books:

In [2]:
books = pd.io.sql.read_sql('''SELECT * FROM books ''', con = engine)


In [3]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Books dataset includes information about 1,000 books, the variable book_id as a primary key and additionaly information about the book- author_id, title, num_pages, publication_date and publisher_id as the publisher's identifier.

### Authors

In [5]:
authors = pd.io.sql.read_sql('''SELECT * FROM authors ''', con = engine)

In [6]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Authors database includes information about 636 authors when author_id is an identifier for authors.

### Publishers

In [8]:
publishers = pd.io.sql.read_sql('''SELECT * FROM publishers''', con = engine)

In [9]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Publishers database includes 340 publishers, the publisher_id is the identifier for purlishers. 

### Ratings

In [11]:
ratings = pd.io.sql.read_sql('''SELECT * FROM ratings''', con = engine)

In [12]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Ratings database includes 6456 ratings, the raters-username, the book_id as identifier for the book being rated and an identifier (rating_id) for every rating

### Reviews

In [14]:
reviews = pd.io.sql.read_sql('''SELECT * FROM reviews''', con = engine)

In [15]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [16]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Reviews database includes written reviews (text) by the users (username), unique identifier for the review (review_id) and book_id identifier.

## Tasks

#### Find the number of books released after January 1, 2000

In [17]:
pd.io.sql.read_sql('''SELECT COUNT(book_id)
                    FROM books 
                    WHERE (publication_date > '2000-01-01')''', con= engine)

count
0    819

There are 819 books that were released after January the 1st, 2000.

#### Find the number of user reviews and the average rating for each book

In [18]:
ratings_reviews= pd.io.sql.read_sql(''' SELECT b.book_id,
                                rev.count_review,
                                rate.avg_rating
                        FROM 
                            (SELECT book_id
                            FROM books) AS b
                        LEFT JOIN 
                            (SELECT book_id,
                                    COUNT(DISTINCT review_id) as count_review
                            FROM reviews
                            GROUP BY book_id) AS rev ON b.book_id=rev.book_id
                        LEFT JOIN
                            (SELECT book_id, 
                                    AVG(rating) as avg_rating
                            FROM ratings
                            GROUP BY book_id) AS rate ON b.book_id=rate.book_id''', con=engine)

In [19]:
ratings_reviews.tail(10)

book_id  count_review  avg_rating
990      997           3.0    3.400000
991      998           4.0    3.200000
992      999           2.0    4.500000
993     1000           4.0    3.833333
994      191           NaN    4.000000
995      672           NaN    5.000000
996       83           NaN    3.666667
997      221           NaN    4.000000
998      387           NaN    4.000000
999      808           NaN    3.000000

In [20]:
ratings_reviews.isna().sum()

book_id         0
count_review    6
avg_rating      0
dtype: int64

In [21]:
ratings_reviews.describe()

book_id  count_review   avg_rating
count  1000.000000    994.000000  1000.000000
mean    500.500000      2.809859     3.898973
std     288.819436      1.055873     0.562376
min       1.000000      1.000000     1.500000
25%     250.750000      2.000000     3.500000
50%     500.500000      3.000000     4.000000
75%     750.250000      3.000000     4.333333
max    1000.000000      7.000000     5.000000

In [22]:
ratings_reviews['count_review'].sum()

2793.0

Overall we see that there are 6 books without any review. The rest 994 books are reviewed 2,793 in total, and 2.8 times in average per book. Regarding ratings, the average score is 3.9.

#### Identify the publisher that has released the greatest number of books with more than 50 pages

In [23]:
pd.io.sql.read_sql('''SELECT p.publisher,
                            COUNT(DISTINCT b.book_id) AS count
                    FROM publishers AS p
                    INNER JOIN books AS b ON p.publisher_id = b.publisher_id
                    WHERE b.num_pages > 50
                    GROUP BY p.publisher
                    ORDER BY count DESC
                    LIMIT 1''', con=engine)

publisher  count
0  Penguin Books     42

The publisher that released the greatest number of books longer than 50 pages is Penguin Books, with 42 books.

#### Identify the author with the highest average book rating
To identify the author we will filter out books with more than 50 ratings

In [27]:
pd.io.sql.read_sql('''SELECT a.author,
                            AVG(r.rating) AS avg_rating
                    FROM authors AS a
                    JOIN books b on b.author_id = a.author_id
                    JOIN ratings AS r ON b.book_id = r.book_id
                    WHERE b.book_id in (SELECT book_id 
                                        FROM ratings
                                        GROUP BY book_id
                                        HAVING COUNT(*) > 49)
                    GROUP BY a.author
                    ORDER BY avg_rating DESC
                    LIMIT 1''', con=engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

J.K. Rowling is the author with the highest average rating for books with more than 50 ratings.

#### Find the average number of text reviews among users who rated more than 50 books

In [25]:
pd.io.sql.read_sql('''SELECT AVG(r.rev_count) AS avg_reviews
                    FROM (SELECT COUNT(text) AS rev_count
                        FROM reviews
                        WHERE username IN (SELECT username
                                            FROM ratings
                                            GROUP BY username
                                            HAVING COUNT(rating_id)>50)
                        GROUP BY username) AS r''', con=engine)

avg_reviews
0    24.333333

Among users who rated 50 or more books, the average number of written reviews is 24 books. 

## Conclusions:
* Out of 1,000 books in total, 819 books were released from 01/01/2000. Meaning that most of the books in the dataset are considerably new.
* Only 6 books weren't reviewed, but all of them were rated.
* Penguin Books is the publisher with the highest number of books longer than 50 pages, having 42 books.
* J.K. Rowling is the author with the highest average rating of 5 points for all her books that were rated more than 50 times.
* Among users who rated 50 books and more, the average written reviews number is 24, meaning that involved users write texts but less than they give a rating to a book.